rfd/0134-cloud-first-deployment.md

```diff
@@ -0,0 +1,203 @@
+reason, the Cloud release and promotion process will be split in 2 parts:
+
+- The workflow that builds artifacts required for the control plane: amd64
+  binaries and Docker images.
+- The workflow that builds remaining artifacts such as deb/rpm packages and
+  Helm charts.
+
+This will allow us to produce rapid releases that affect only control plane
+services (auth, proxy) while still having the ability to push remaining artifacts
+when needed.
+
+Both workflows will run in Github Actions and will reuse existing bits of the
+full self-hosted release pipeline.
+
+### Cloud promotion process
+
+The Cloud release promotion process will be staged, with a different class of
+tenants being upgraded at each stage:
+
+1. Canary
+2. Internal
+3. Trials
+4. Team & Enterprise
+
+The canary tenants will be a set of precreated clusters with some data in them
+which engineers will use for connecting their agents and testing their features.
+Canary clusters will be updated automatically once a Cloud release is cut.
+
+Before promoting the release further, engineers will test any significant change
+in the cloud staging environment and then explicitly sign off that the change is
+ready for deployment by applying a label (e.g. `cloud/verified`) on the master
+PR corresponding to the change. The release manager then will approve the
+further rollout.
+
+Once the change is rolled out to production tenants, engineers will be expected
+to monitor error rates and Cloud metrics relevant to their changes.
+
+Let's now look at some of the edge case scenarios in the release process and
+how we'll be handling them.
+
+### Behavior-incompatible changes
+
+Teleport's component compatibility guarantees become somewhat obsolete in a
+Cloud-first world. Historically, we've been reserving behavior-incompatible
+changes for major releases. The concept of a major release, however, becomes
+non-existent in the Cloud-first world where any change, backwards compatible
+or not, gets deployed to all users within days.
+
+When applied to Cloud release model, the compatibility requirement becomes:
+
+"Any Cloud release must be always compatible with the previous Cloud release"
+
+Compatibility requirements affect control plane (proxy, auth and other services
+we host for our users), agents and clients:
+
+- Control plane. We have full control over it, deploy in tandem and make sure
+  these are always compatible with each other.
+- Agents deployed by users. We always keep backwards compatibilithy between
+  adjacent releases and rely on automatic upgrades to make sure they don't drift.
+- Clients. Clients require automatic upgrades to avoid breakage on behavior
+  incompatible changes. This is [in progress](https://github.com/gravitational/cloud/issues/4880).
+
+### Rollbacks
+
+Bad releases can happen even with the multi-staged rollout process. In that
+situation the process will be:
+
+- For the control plane, we will prefer rolling forward in most cases to a new
```

Comment:
> we will prefer rolling forward

A word of caution, this goes against some of the more solid oncall practices I've seen elsewhere. If a rollback would solve the situation, say by avoiding a newly-added bad behavior, then it's one of the first actions I'd expect a oncall to take. It'll not be easy to get us into a state of where rollbacks are mainly safe, but it is doable and a worthy objective to have.

I don't expect changes to the design doc, as I'm sure there's not only one way of doing things, but I think this is worth saying nonetheless.

Reply:
How will we manage restoring migrated resources in the event of a rollback? Do we have to look at having more sophisticated tooling for managing our resource schemas than just upgrade-on-first-run? Is this part of the reason for favouring a "roll-forward" approach?

Reply:
Trunk-based CD is inherently different from our milestone-based approach. We'll push much more frequently, meaning we shouldn't hit as many migrations as folks do when bumping majors, for example.

That said, we'll have to be more careful with migrations, for sure. I won't go about listing alternatives here, lest the thread becomes and endless parade of what-ifs; instead I'll say this is a known problem with many viable solutions.

> Is this part of the reason for favouring a "roll-forward" approach?

I suppose so? It's a weak argument, I think. Leaving our customers without ways to rollback is just as bad as removing to option to ourselves.

Reply:
I don't think we really intend for this to be a hard prescription, "we only roll forward". It is more of a guidance I guess: if our Cloud release pipeline is quick and, say, the issue only affects the control plane, it might be cheaper/faster/less risky for us to just fix the issue, push a new release, upgrade Cloud and trigger auto-upgrades, than doing rollbacks. I think in practice we'll just decide on a case-by-case basis.